aws-sdk-core/lib/aws-sdk-core/signers/v4.rb

```diff
@@ -39,10 +39,16 @@ def self.sign(context)
       #   the endpoint prefix.
       # @param [String] region The region (e.g. 'us-west-1') the request
       #   will be made to.
-      def initialize(credentials, service_name, region)
+      # @param [Array] whitelist headers provided to escape blacklist check
+      def initialize(credentials, service_name, region, whitelist_headers = [])
         @service_name = service_name
         @credentials = credentials.credentials
         @region = EndpointProvider.signing_region(region, service_name)
+        if whitelist_headers.any? && (BLACKLIST_HEADERS & whitelist_headers).any?
```

Comment:
I might be missing something, but why not just remove this whole block and use:
```ruby
@blacklist = BLACKLIST_HEADERS - whitelist_headers
```
You already set whitelist headers to be an array by default, so once we get to this point, I'm not sure the if-else block is doing anything.